### ISSUE TITLE
How do I make nemo log only on rank 0?

### ISSUE THREAD

**Issue by stas00** [NVIDIA/NeMo#7495]
I can't figure out how to make the logging not repeat on 512 ranks. Is there a setting to tell nemo to log only on rank 0? at the very least for the iteration logs?

I currently have:
```
  exp_manager.create_dllogger_logger: True
```
but I'd happy with any other way.

What I ideally would like is the original Megatron-LM-like style logging where it logs iteration metrics from rank 0.

Thank you!

**Comment by titu1994** [comment]
Yup you can use this flag in Exp Manager - https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/exp_manager.py#L166-L167

Either 1 per local rank 0, or just 1 for global rank 0 in total.

**Comment by stas00** [comment]
oh, I already had:

```
exp_manager:
[...]
  log_local_rank_0_only: True
```

but it doesn't help, I get logging on each rank

**Comment by titu1994** [comment]
Odd, this should be fine - https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/exp_manager.py#L393-L402

**Comment by stas00** [comment]
this looks like a different logger to a specific exp_manager file, I'm talking about what goes to `stdout`

I'd imagine it's because of `exp_manager.create_dllogger_logger: True` but I have no idea how to control its behavior as there are no such config options that I can see.

**Comment by titu1994** [comment]
Dllogger doesn't have that many fine grained options. If they do enable it, please do add a PR to add those arguments !

**Comment by stas00** [comment]
Hmm, are you sure this is even in the Dllogger land? Its API appears to be totally unaware of ranks.

Should it be something like this on nemo side instead?
```
if (log_local_rank_0_only and rank==0) or not log_local_rank_0_only:
  DLLogger.log(...)
```
using pseudo-code as I don't yet know the actual variables you use.

**Comment by stas00** [comment]
As @jan-golda confirmed here: https://github.com/NVIDIA/dllogger/issues/9#issuecomment-1734541555
this is indeed a job for NeMo.

**Comment by titu1994** [comment]
NeMo does not special case loggers, we get the implementation from here - https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/loggers/dllogger.py

If you want, you can submit a pr to add a check here to only log on rank 0. But the proposed solution isn't a general one so it can't be used.

**Comment by stas00** [comment]
Could you please why?

I see that you have this call:

https://github.com/NVIDIA/NeMo/blob/af34a958c484a2af76ed855cdee9a3c6bebd6743/nemo/collections/common/callbacks/callbacks.py#L34

and it's this call that is not respecting the `log_local_rank_0_only` setting if I understand things correctly.

So the fix is not logger specific but a global one that should do the right thing for any logger.

Am I wrong?

**Comment by stas00** [comment]
Alternatively, may be if `log_local_rank_0_only` is true, then the logger shouldn't be initialized on any but rank0 processes and init to some sort of DevNull logger for all other ranks - that way no code will need to be changed ever instead of hunting down all the many instances where logging is called.

this is probably a much saner solution, no?

**Comment by titu1994** [comment]
All other loggers properly handle logging on multiple ranks, some even use this functionality such as wandb to plot and collate results for gpu from multiple ranks. 

What you can do is put the DLLogger.log_* methods under a ptl rank_zero_only() decorator so that all other ranks ignore it. 

Log local rank 0 only is just for the NeMo log files 
written to by calling nemo.utils.logging.*. It has no effect on any other part of the experiment being run, including calls to trainer.logger().

**Comment by stas00** [comment]
Thank you for clarifying, Somshubra - the reason for sending to loggers on all ranks makes sense.

But I have no control over when `DLLogger.log_*` methods are called - this is all inside Nemo.

If I put `@rank_zero_only` decorator from PL on the `main` function of the program the whole thing hangs `Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/8` - it warns in PL doc that this decorator can lead to deadlocks and requires manual synchronization. So most likely it's meant to be used at the point of logging. reference: https://lightning.ai/docs/pytorch/stable/extensions/logging.html

Honestly, I don't even understand how this is meant to work if it says at the logging doc above:
> rank_zero_only: Whether the value will be logged only on rank 0. This will prevent synchronization which would produce a deadlock as not all processes would perform this log call.

I suppose it means that whoever uses this flag needs to manually call a `barrier` right after the call, but that's not helping us to resolve this Issue.

**Comment by titu1994** [comment]
Don't put that decorator in the main function, put it here - https://github.com/NVIDIA/NeMo/blob/main/nemo/utils/loggers/dllogger.py#L84

**Comment by stas00** [comment]
That did the trick, thank you very much for your help and the patience, @titu1994!

The PR is here: https://github.com/NVIDIA/NeMo/pull/7513

I don't know if it should also be documented that Dllogger will log on rank 0?

**Comment by titu1994** [comment]
Great ! Could you add a comment to the init docstring stating that it uses only rank 0 for logging ?
